raspiraw.c

```diff
@@ -695,6 +861,80 @@ int main(int argc, const char** argv) {
 		return -2;
 	}
 
+
+	if (cfg.regs)
+	{
+		int r,b;
+		char *p,*q;
+
+		p=strtok(cfg.regs, ";");
+		while (p)
+		{
+			vcos_assert(strlen(p)>6);
+			vcos_assert(p[4]==',');
+			vcos_assert(strlen(p)%2);
+			p[4]='\0'; q=p+5;
+			sscanf(p,"%4x",&r);
+			while(*q)
+			{
+				vcos_assert(isxdigit(q[0]));
+				vcos_assert(isxdigit(q[1]));
+
+				sscanf(q,"%2x",&b);	
+				vcos_log_error("%04x: %02x",r,b);    
+
+				set_reg(sensor_mode, r, b);
+
+				++r;
+				q+=2;
+			}
+			p=strtok(NULL,";");
+		}
+	}
+
+	if (cfg.hinc >= 0)
+	{
+		// TODO: handle modes different to ov5647 as well
+		set_reg(sensor_mode, 0x3814, cfg.hinc);
+	}
+
+	if (cfg.vinc >= 0)
+	{
+		// TODO: handle modes different to ov5647 as well
+		set_reg(sensor_mode, 0x3815, cfg.vinc);
+	}
+
+	if (cfg.fps > 0)
+	{
+		// TODO: handle modes different to ov5647 as well
+		int h = get_reg(sensor_mode, 0x380E);
+		int l = get_reg(sensor_mode, 0x380F);
+                int n = ((h<<8) + l) * sensor_mode->framerate / cfg.fps;
+		set_reg(sensor_mode, 0x380E, n>>8);
+		set_reg(sensor_mode, 0x380F, n&0xFF);
+	}
+
+	if (cfg.width > 0)
+	{
+		sensor_mode->width = cfg.width;
+	}
+
+	if (cfg.height > 0)
+	{
+		sensor_mode->height = cfg.height;
+	}
+
```

Comment:
Thinking aloud more than anything.
To alter the crop you're currently forcing the user to specify register values and then width/height - lots of potential for errors, and I have seen it result in memory corruption (I wish I knew exactly why, as it shouldn't be possible).

Are we better off with left/right/top/bottom crop parameters that then compute the registers and width / height? I'm happy enough with a function pointer in the sensor_def to a function that works out the values rather trying to write generic code with lookup tables of registers.
hinc and vinc are the awkward ones, but again messing with those really ought to be done as a new mode.